### PR TITLE
Add weather provider integration support (B.18)

### PR DESCRIPTION
**Pull request by Pierre-Gilles**
### Description

Implements phase 2 workstream B.18: external weather integrations. This allows third-party weather providers to be installed as external integrations and automatically takes precedence over the built-in OpenWeather service.

**Key changes:**

1. **Weather provider interface**: External integrations with `type: 'weather'` in their manifest now expose a `weather.get(options)` capability that mirrors the internal OpenWeather service interface, enabling duck-typed provider switching.

2. **Payload normalization**: Added `externalIntegration.normalizeWeather()` that validates and bounds weather payloads from unaudited external code:
   - Whitelists all fields (unknown fields dropped)
   - Coerces numbers to finite values, dates to valid Date objects
   - Caps arrays (24 hours, 8 days, 10 alerts) and string lengths
   - Coerces unknown weather conditions to 'unknown'
   - Enforces CAP-style alert severities (minor/moderate/severe/extreme)

3. **Provider selection logic**: Updated `weather.get()` to iterate through all available weather providers (external + internal) sorted by name, with `ext-*` selectors taking precedence over `openweather`. Falls back gracefully when a provider fails, distinguishing between "not configured" and real errors.

4. **UI enhancements**:
   - Weather box now displays alerts with severity-based color coding (blue/yellow/orange/red)
   - Shows sunrise/sunset times and UV index when available
   - Updated error messages to reference generic weather configuration rather than OpenWeather specifically
   - Added weather icon mappings for drizzle and thunderstorm conditions

5. **Manifest schema**: Extended to accept `type: 'weather'` for external integrations.

6. **WebSocket protocol**: Added `WEATHER_GET` message type with 15s timeout (matching camera image freshness requirement).

All changes are covered by comprehensive unit tests validating normalization edge cases, provider fallback behavior, and integration with the proxy service system.

### Checklist

- [x] Tests pass: Added 262 lines of test coverage for weather normalization and provider selection
- [x] Linter and prettier pass
- [x] No breaking changes: Weather provider selection is backward compatible; existing OpenWeather setups continue to work

https://claude.ai/code/session_01FuafgHvr5244RxZ5KKTN3a

## Summary by CodeRabbit

* **New Features**
  * Added support for weather-provider integrations alongside device and communication integrations.
  * Weather retrieval supports provider fallback and location, language, and unit options.
  * Enhanced weather dashboards with alerts, sunrise/sunset times, humidity, wind, UV index, and additional forecast icons.
  * Added weather-specific installation guidance and localized messaging.

* **Bug Fixes**
  * Improved handling of incomplete forecasts, unavailable icons, future dates, and missing weather details.
  * Added validation and limits for weather data to support reliable display.

### CODE REVIEW

File: server/lib/weather/weather.get.js [hunk 1]
```diff
@@ -18,12 +30,48 @@ const { ServiceNotConfiguredError } = require('../../utils/coreErrors');
...
+  const serviceNames = this.service.stateManager.getAllKeys('service');
+  const candidates = serviceNames
+    .filter((serviceName) => {
+      const service = this.service.getService(serviceName);
+      return service && service.weather && typeof service.weather.get === 'function';
+    })
+    .sort();
+  if (candidates.length === 0) {
+    throw new ServiceNotConfiguredError('No weather provider is installed or configured.');
+  }
```
[the remark this answers is not in the document]
Comment (Pierre-Gilles) [reply server/lib/weather/weather.get.js:53]: Declining this one — the premise about the timeout doesn't hold:

- **A stopped/disconnected integration never incurs the 15 s timeout.** `sendCommand` checks the WebSocket connection map first and throws `EXTERNAL_INTEGRATION_NOT_CONNECTED` immediately when the socket is not open (`externalIntegration.sendCommand.js:24-27`) — the timer is only armed for a connected integration. So a stopped external weather provider falls through to `openweather` instantly; that fall-through *is* the designed mechanism, same as the `message.*` capability which also stays registered while a communication integration is stopped.
- The 15 s deadline can only be paid for a *connected but unresponsive* integration — which is exactly the case where trying it is legitimate (it may answer), and the deadline is the bound.
- On the stale-state concern: the candidate filter duck-types `service.weather.get` and `weather.get` re-resolves via `getService` inside the try, so a `null` service is simply skipped/caught and the loop moves on.

Deregistering the proxy on stop would create the asymmetry the proxy design deliberately avoids (proxies are registered once per `t_service` row, lifecycle is expressed through connection state), for no latency gain.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/lib/weather/weather.checkAlerts.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      });
+    } catch (e) {
+      // no provider configured or provider down: nothing to diff, the
+      // previous baseline is kept so recovery does not re-fire scenes
+      logger.debug(`weather.checkAlerts: no weather for house ${house.selector}: ${e.message}`);
+      return;
+    }
+    const alerts = weather.alerts || [];
+    const previous = this.houseAlerts.get(house.selector);
+    this.houseAlerts.set(house.selector, alerts);
```
[the remark this answers is not in the document]
Comment (Pierre-Gilles) [reply server/lib/weather/weather.checkAlerts.js:68]: Fixed in d6a58ac: `checkAlerts` now takes an in-flight guard — a check landing while another is running is dropped immediately (a dropped nudge costs at most the 30-min floor). Covered by a test that starts an overlapping check against a hanging provider and asserts a single provider call and a single raised event, plus a follow-up check proving the guard is released. Spec updated (B.18 point 5).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/lib/weather/weather.command.js [hunk 3]
```diff
@@ -78,6 +80,14 @@ async function command(message, classification, context) {
             context.day = capitalizeFirstLetter(dateEntity.sourceText);
```
[the remark this answers is not in the document]
Comment (Pierre-Gilles) [reply server/lib/weather/weather.command.js:81]: Fixed in d6a58ac: the day condition falls back to the `unknown` templates (`weather.get.success.<day>.unknown` exists in both en and fr answer files), so a provider omitting the per-day condition — and OpenWeather's own day entries — still answers with the min/max temperatures. Test added.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/api/controllers/weather.controller.js [hunk 4]
```diff
@@ -61,8 +61,26 @@ module.exports = function WeatherController(gladys) {
...
+   * @apiSuccessExample {json} Success-Example
+   * {
+   *   "image": "data:image/png;base64,iVBORw0KGgo..."
+   * }
+   */
+  async function getImage(req, res) {
+    // resolves the house first: the route stays scoped like the weather
+    // route it extends, and an unknown house 404s before any provider call
+    await gladys.house.getBySelector(req.params.house_selector);
+    const image = await gladys.weather.getImage(req.params.image_key);
```
[the remark this answers is not in the document]
Comment (Pierre-Gilles) [reply server/api/controllers/weather.controller.js]: Documented in d6a58ac rather than threaded: B.18 point 6 now states that provider images are instance-global, not per-house — the house selector on the route only scopes authentication/404, and `weather.get-image` deliberately carries no coordinates. The v1 use cases (national vigilance map, radar of the area the provider already serves) don't need per-house images, and adding house coordinates to the command later is a purely additive extension if the need materializes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/lib/external-integration/externalIntegration.normalizeWeatherImage.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * @param {any} rawBase64 - The data.image of the command-result.
+ * @returns {string} The validated image as a data URI.
+ * @example
+ * const image = normalizeWeatherImage('iVBORw0KGgo...');
+ */
+function normalizeWeatherImage(rawBase64) {
+  if (typeof rawBase64 !== 'string' || rawBase64.length === 0) {
+    throw new ExternalIntegrationUnavailableError('EXTERNAL_INTEGRATION_INVALID_WEATHER_IMAGE');
+  }
+  const bytes = Buffer.from(rawBase64, 'base64');
```
[the remark this answers is not in the document]
Comment (Pierre-Gilles) [reply server/lib/external-integration/externalIntegration.normalizeWeatherImage.js:28]: Fixed in d6a58ac: base64 strings longer than `ceil(MAX_WEATHER_IMAGE_BYTES / 3) * 4 + 4` are rejected before `Buffer.from` allocates anything; the decoded-size check stays as the exact bound (a string can pass the length pre-check and still decode a few bytes over the cap). Both rejection paths are covered by tests.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/lib/external-integration/externalIntegration.registerProxyService.js [hunk 6]
```diff
@@ -77,6 +84,54 @@ function registerProxyService(service) {
...
+              return cached.image;
+            }
+            const result = await this.sendCommand(
+              service,
+              WEBSOCKET_MESSAGE_TYPES.EXTERNAL_INTEGRATION.WEATHER_GET_IMAGE,
+              { key },
+              { timeoutMs: WEATHER_GET_TIMEOUT_MS },
+            );
+            const image = normalizeWeatherImage(result && result.data && result.data.image);
+            this.cache.set(cacheKey, { image, expiresAt: Date.now() + WEATHER_IMAGE_CACHE_TTL_MS });
```
[the remark this answers is not in the document]
Comment (Pierre-Gilles) [reply server/lib/external-integration/externalIntegration.registerProxyService.js:143]: Fixed in 4634aba with a two-layer allowlist:

- The proxy now only relays keys **declared in the last normalized weather payload** of that integration — an undeclared key throws `NotFoundError` before a single byte reaches the integration, so cache cardinality is bounded by the existing 3-image cap per integration (and the provider fetch load for declared keys was already bounded by the 10-min cache).
- `weather.getImage` additionally shape-checks the requested key against the declaration regex (`^[a-z0-9][a-z0-9-]{0,31}$`) before consulting any provider, so an attacker-sized string never enters a cache key, a log line or an error message.

Spec updated (B.18 point 6 + C.4), covered by tests (undeclared key refused with zero commands sent, re-declaration semantics, shape rejection).

---
_Generated by [Claude Code](https://claude.ai/code)_
